src/Interpreters/AsynchronousInsertQueue.cpp

```diff
@@ -429,17 +567,21 @@ void AsynchronousInsertQueue::processBatchDeadlines(size_t shard_num)
         {
             std::unique_lock lock(shard.mutex);
 
-            shard.are_tasks_available.wait_for(lock,
-                Milliseconds(getContext()->getSettingsRef().async_insert_busy_timeout_ms), [&shard, this]
-            {
-                if (shutdown)
-                    return true;
+            const auto rel_time
+                = std::min(shard.busy_timeout_ms, Milliseconds(getContext()->getSettingsRef().async_insert_poll_timeout_ms));
+            shard.are_tasks_available.wait_for(
+                lock,
+                rel_time,
```

Comment:
That's how it's implemented right now. Imagine an entry arriving in the queue just after a timeout has fired. When the timeout fires again, that entry won't be processed. Instead, it will stay in the queue until the next cycle. Essentially, that entry will remain in the queue for almost double the intended time.
If entries arrive uniformly over time, they are expected to remain in the queue for an additional half of the timeout.
This is especially visible for large (over 5s) timeouts and frequent inserts.

Reply:
... e.g., the following tests:
```
find tests/ -name '*.sh' -o -name '*.sql' 2>/dev/null | xargs grep 'async_insert_busy_timeout.*00000'
tests/queries/0_stateless/02015_async_inserts_2.sh:url="${CLICKHOUSE_URL}&async_insert=1&wait_for_async_insert=1&async_insert_busy_timeout_ms=600000&async_insert_max_query_number=3&async_insert_deduplicate=1"
tests/queries/0_stateless/02726_async_insert_flush_queue.sql:SET async_insert_busy_timeout_ms = 1000000;
tests/queries/0_stateless/02884_async_insert_native_protocol_1.sh:async_insert_options="--async_insert 1 --wait_for_async_insert 0 --async_insert_busy_timeout_ms 1000000"
tests/queries/0_stateless/02884_async_insert_native_protocol_3.sh:async_insert_options="--async_insert 1 --wait_for_async_insert 0 --async_insert_busy_timeout_ms 1000000"
tests/queries/0_stateless/02884_async_insert_skip_settings.sql:SET async_insert_busy_timeout_ms = 100000;
```
The reason why they don't fail is that the condvar polling time is taken from the server settings (default 200ms), not the query settings.

Reply:
okay, that makes sense, thanks for explaining

Reply:
Instead of introducing an additional `async_insert_poll_timeout_ms` setting, we may use a fixed fraction (e.g., 1/8) of `busy_timeout_ms`.

Reply:
no, I think it will make it worse because tuning it will be abstract
I think polling makes sense and we can leave the setting so you can always fine tune it